crates/accelerate/src/qi/versor_u2.rs

```diff
@@ -0,0 +1,547 @@
+        self.matrix_contiguous_into(&mut out);
+        out
+    }
+}
+impl From<VersorSU2> for VersorU2 {
+    fn from(val: VersorSU2) -> Self {
+        Self {
+            phase: 0.,
+            su2: val,
+        }
+    }
+}
+
+/// Implement the `Mul` traits between two `Copy` types for the pairs `[(&T, U), (T, &U), (&T, &U)]`
+/// by delegating the reference-based multiplications to the owned-based version using `Copy`.
+macro_rules! impl_mul_refs {
+    ($right:ty, $left:ty) => {
+        impl ::std::ops::Mul<$left> for &$right {
+            type Output = <$right as ::std::ops::Mul<$left>>::Output;
+            fn mul(self, other: $left) -> Self::Output {
+                *self * other
+            }
+        }
+        impl ::std::ops::Mul<&$left> for $right {
+            type Output = <$right as ::std::ops::Mul<$left>>::Output;
+            fn mul(self, other: &$left) -> Self::Output {
+                self * *other
+            }
+        }
+        impl ::std::ops::Mul<&$left> for &$right {
+            type Output = <$right as ::std::ops::Mul<$left>>::Output;
+            fn mul(self, other: &$left) -> Self::Output {
+                *self * *other
+            }
+        }
+    };
+}
+
+impl ::std::ops::Mul for VersorSU2 {
+    type Output = VersorSU2;
+    fn mul(self, other: VersorSU2) -> Self::Output {
+        VersorSU2(self.0 * other.0)
+    }
+}
+impl_mul_refs!(VersorSU2, VersorSU2);
+
+impl ::std::ops::Mul for VersorU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorU2) -> Self::Output {
+        VersorU2 {
+            phase: self.phase + other.phase,
+            su2: self.su2 * other.su2,
+        }
+    }
+}
+impl_mul_refs!(VersorU2, VersorU2);
+
+impl ::std::ops::Mul<VersorU2> for VersorSU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorU2) -> VersorU2 {
+        VersorU2 {
+            phase: other.phase,
+            su2: self * other.su2,
+        }
+    }
+}
+impl_mul_refs!(VersorU2, VersorSU2);
+
+impl ::std::ops::Mul<VersorSU2> for VersorU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorSU2) -> VersorU2 {
+        VersorU2 {
+            phase: self.phase,
+            su2: self.su2 * other,
+        }
+    }
+}
+impl_mul_refs!(VersorSU2, VersorU2);
+
+/// A module-internal trait to simplify the code-generation of both the dynamic `ndarray` and the
+/// static `&[[Complex64; 2]; 2]` and dynamic `ndarray` paths.  Rather than making the user care
+/// about importing it, we just expose the concretised methods using it through `VersorU2`.
+trait Matrix1q {
+    fn get(&self, row: usize, col: usize) -> Complex64;
+}
+impl<const N: usize, const M: usize> Matrix1q for [[Complex64; N]; M] {
+    #[inline(always)]
+    fn get(&self, row: usize, col: usize) -> Complex64 {
+        self[row][col]
+    }
+}
+impl Matrix1q for ArrayView2<'_, Complex64> {
+    #[inline(always)]
+    fn get(&self, row: usize, col: usize) -> Complex64 {
+        self[(row, col)]
+    }
+}
```

Comment:
Do you want to add a `Matrix2` impl here too? I guess nothing needs it now because you only used static arrays and ndarray here. But I feel now that we have `UnitaryGate` that can be built from a `Matrix2<Complex64>` it might be good to have the impl here so we can call `VersorU2::from_matrix_unchecked` directly on that.

Reply:
Yeah, I'll update this in the face of the new `UnitaryGate` for sure.

Reply:
Done in 3f20b0fe75.